### PR TITLE
Refactor Modules and Add CLI

### LINKED ISSUE #8: Increase User Friendliness of Scenario When the Wrong App Name is Passed In
Currently if the factory is configured with the wrong application name and the plugin can't find the app in the json config, we get a NPE.

We should consider the following:
1. Check for NPE and throw user friendly exception
2. Consider checking for application name without case sensitivity

### PR DESCRIPTION
**Pull request by sherl0cks**
#### What does this PR do?
- moves generation code to a core module
- separates out jenkins plugin to its own module
- adds a java cli module
- improves error handling
- add Jenkinsfile for CI

#### Which tests illustrate how this code works?
see change set

#### Is there a relevant Issue open for this?
resolves #8 

#### Who would you like to review this?
/cc @mdanter

### CODE REVIEW

File: jenkins-pipeline-generator-cli/pom.xml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        <license>
+            <name>ASL 2.0</name>
+            <url>https://opensource.org/licenses/Apache-2.0</url>
+        </license>
+    </licenses>
+    <dependencies>
+        <dependency>
+            <groupId>com.rhc</groupId>
+            <artifactId>jenkins-pipeline-generator-core</artifactId>
+            <version>${project.version}</version>
```
Comment (mdanter) [inline jenkins-pipeline-generator-cli/pom.xml]: please use the parent pom to define a dependency management section for the libraries with corresponding versions.

File: jenkins-pipeline-generator-core/pom.xml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>com.rhc</groupId>
+            <artifactId>automation-api-client</artifactId>
+            <version>${labs.api.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+            <version>1.7.22</version>
```
Comment (mdanter) [inline jenkins-pipeline-generator-core/pom.xml]: reiterating for sake of being pedantic, please move these into the parent's dependencyManagement section.

File: jenkins-pipeline-generator-cli/src/test/java/com/rhc/jenkinsfile/generator/cli/JenkinsfileGeneratorCliTest.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package com.rhc.jenkinsfile.generator.cli;
```
Comment (mdanter) [inline jenkins-pipeline-generator-cli/src/test/java/com/rhc/jenkinsfile/generator/cli/JenkinsfileGeneratorCliTest.java]: some of the files have a license header, some don't. Could you make this consistent?

File: jenkins-pipeline-generator-core/src/main/java/com/rhc/jenkinsfile/generator/EngagementMarshaller.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    public static Engagement getEngagementFromFileOnClasspath(String fileName){
+        InputStream is = EngagementMarshaller.class.getClassLoader().getResourceAsStream(fileName);
+        if (is == null) {
+            throw new RuntimeException("Could not find the specified file: " + fileName);
+        }
+        Engagement engagement = getEngagementFromInputStream(is);
+        return engagement;
+    }
+
+    public static Engagement getEngagementFromFileInWorkingDirectory(String fileName){
```
Comment (mdanter) [inline jenkins-pipeline-generator-core/src/main/java/com/rhc/jenkinsfile/generator/EngagementMarshaller.java]: some of your code was using final on args, some isn't, this is a total nit-pick

File: jenkins-pipeline-generator-plugin/pom.xml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            <artifactId>workflow-cps</artifactId>
+            <version>${jenkins.cps.version}</version>
+        </dependency>
+
+
+        <!-- Jenkins Pipe test harness -->
+        <dependency>
+            <groupId>org.jenkins-ci.plugins.workflow</groupId>
+            <artifactId>workflow-basic-steps</artifactId>
+            <version>${jenkins.cps.version}</version>
```
Comment (mdanter) [inline jenkins-pipeline-generator-plugin/pom.xml]: dependencyManagement in the parent pom for version

File: pom.xml [hunk 6]
```diff
@@ -1,139 +1,40 @@
...
+
+    <name>Jenkins Pipeline Generator :: Parent</name>
+    <url>https://github.com/rht-labs/dynamic-pipeline-jenkins-plugin</url>
+
+    <properties>
+        <labs.api.version>0.3.0-SNAPSHOT</labs.api.version>
+        <logback.version>1.1.7</logback.version>
+        <mockito.version>2.1.0-RC.1</mockito.version>
+    </properties>
+
```
Comment (mdanter) [inline pom.xml]: put in a depedencyManagement section with versions of libs being referenced from properties

File: jenkins-pipeline-generator-plugin/pom.xml (hunk 5 above)
Comment (sherl0cks) [inline jenkins-pipeline-generator-plugin/pom.xml]: per chat, this isn't possible with jenkins plugin as it needs to use the Jenkins parent pom

**Comment by sherl0cks** [comment]
@mdanter all changes that can be made have been pushed

**Pull request merged**